src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +649,125 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
 
 
 /*
- * TableReferencing function checks whether given table is referencing by another table
- * via foreign constraints. If it is referencing, this function returns true. To check
- * that, this function searches given relation at pg_constraints system catalog. However
- * since there is no index for the column we searched, this function performs sequential
- * search, therefore call this function with caution.
+ * TableReferencing function checks whether given table is referenced to another
+ * table via foreign key constraints. If it is referencing, this function returns
+ * true.
  */
 bool
 TableReferencing(Oid relationId)
 {
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	Oid scanIndexId = InvalidOid;
-	bool useIndex = false;
-
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+	bool excludeSelfReference = false;
+	bool findAll = false;
+	return GetTableForeignConstraintOidsInternal(relationId, Anum_pg_constraint_conrelid,
+												 excludeSelfReference, findAll) != NIL;
+}
 
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, scanIndexId, useIndex,
-													NULL,
-													scanKeyCount, scanKey);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+/*
+ * ConstraintIsAForeignKey returns true if the given constraint name
+ * is a foreign key defined on the relation.
+ */
+bool
+ConstraintIsAForeignKey(char *constraintNameInput, Oid relationOid)
+{
+	bool excludeSelfReference = false;
+	bool findAll = true;
+	List *foreignKeys = GetTableForeignConstraintOidsInternal(relationOid,
+															  Anum_pg_constraint_conrelid,
+															  excludeSelfReference,
+															  findAll);
+
+	Oid tableForeignConstraintOid = InvalidOid;
+	foreach_oid(tableForeignConstraintOid, foreignKeys)
```

Comment:
So here we check if a constraint name exists in a list of constraints. I think we were doing the same thing in `ConstraintIsAForeignKeyToReferenceTable`, so we could create a method that takes a list and a constraint name to check if constraint name exists in that list.

Reply:
makes sense